lib/components/user/notification-prefs-pane.js

```diff
@@ -218,68 +155,244 @@ class NotificationPrefsPane extends Component {
+
+    // Empty the input box (and disable the Send text button).
+    setFieldValue('newPhoneNumber', '')
+    setFieldTouched('newPhoneNumber', false)
+
+    const submittedNumber = new PhoneNumber(newPhoneNumber, phoneFormatOptions.regionCode).getNumber('e164')
+    const result = await onRequestCode(submittedNumber)
+
+    if (result && result.phoneNumber === submittedNumber && !result.isPhoneNumberVerified) {
+      // On success (phoneNumber is updated and verified status is false), update the formatted phone number in Formik's state.
+      setFieldValue('newPhoneNumber', formatPhoneNumber(result.phoneNumber, phoneFormatOptions))
+
+      // Set empty code field in Formik's state and prompt for code.
+      setFieldValue('validationCode', '')
+      setFieldTouched('validationCode', false)
+      this.setState({
+        isEditing: false,
+        isPending: true
+      })
+    }
+  }
+
+  _handleSubmitCode = async () => {
+    const { onSubmitCode, onVerified, phoneFormatOptions, setFieldValue, values } = this.props
+    const { validationCode } = values
+
+    // Clear the code field and disable the verify button.
+    // (The code to send is already captured in const above.)
+    setFieldValue('validationCode', '')
+
+    const result = await onSubmitCode(validationCode)
+
+    if (result && result.isPhoneNumberVerified) {
+      // On success (status is changed to verified), update the phone number in Formik's state and hide the code field.
+      setFieldValue('newPhoneNumber', '')
+
+      this.setState({
+        currentNumber: formatPhoneNumber(result.phoneNumber, phoneFormatOptions),
+        currentNumberVerified: true,
+        isPending: false
+      })
+
+      onVerified(result)
+    }
+  }
+
+  render () {
+    // All props below are Formik props (https://formik.org/docs/api/formik#props-1)
+    // except where indicated.
+    const {
+      errors,
+      phoneFormatOptions, // Provided by parent.
+      touched,
+      values
+    } = this.props
+    const { newPhoneNumber } = values
+    const { currentNumber, currentNumberVerified, isEditing, isPending } = this.state
+
+    // Here are the states we are dealing with:
+    // - First time entering a phone number/validation code (blank value, not modified)
+    //   => no color, no feedback indication.
+    // - Typing backspace all the way to erase a number/code (blank value, modified)
+    //   => red error.
+    // - Typing a phone number that doesn't match the configured phoneNumberRegEx
+    //   => red error.
+    const phoneErrorState = getErrorState(this.props, 'newPhoneNumber')
+    const codeErrorState = getErrorState(this.props, 'validationCode')
+
+    return (
+      <>
+        {isEditing
+          ? (
+            <FormGroup validationState={phoneErrorState}>
+              <ControlLabel>Enter your phone number for SMS notifications:</ControlLabel>
+              <ControlStrip>
+                <Field as={InlineTextInput}
```

Comment:
As mentioned above, we should just use https://catamphetamine.gitlab.io/react-phone-number-input/